### ISSUE TITLE
CORS is not working on strapi@4.24.2

### ISSUE THREAD

**Issue by bugmaster99** [strapi/strapi#20672]
### Required System information

- Node.js version: v18.12.0
- NPM version: npm/8.19.2
- Strapi version: 4.24.2
- Database: mysql
- Operating system: macOS :- darwin-arm64
- Is your project Javascript or Typescript: Javascript

### Describe the bug

i'm trying to add CORS for some of my sub-domains/domains such that only those will be able to access my strapi resources but unfortunately i'm unable to do so. I followed the documentation as per https://docs.strapi.io/dev-docs/configurations/middlewares, still no luck, I can still access strapi backend through any domain (currently testing with http://localhost:3000/ - my nextjs project (strapi running on http://localhost:1337)). I tried adding @koa/cors and added in resolutions which was mentioned in some other issue similar to mine but still no luck.

### Expected behavior

Expected behaviour should be that my strapi content should only be accessed by those origins which are in the list (origin list in cors config setting)

### Code snippets

strapi package.json :-
```
{
  "name": "blog-cms",
  "version": "0.1.0",
  "private": true,
  "description": "A Strapi application",
  "license": "MIT",
  "author": {
    "name": "A Strapi developer"
  },
  "scripts": {
    "build": "strapi build",
    "dev": "strapi develop",
    "start": "strapi start",
    "strapi": "strapi"
  },
  "resolutions": {
    "@koa/cors": "5.0.0"
  },
  "dependencies": {
    "@_sh/strapi-plugin-ckeditor": "^1.1.1",
    "@aws-sdk/client-secrets-manager": "^3.480.0",
    "@koa/cors": "5.0.0",
    "@strapi/plugin-color-picker": "^4.9.1",
    "@strapi/plugin-graphql": "^4.24.2",
    "@strapi/plugin-i18n": "4.24.2",
    "@strapi/plugin-users-permissions": "4.24.2",
    "@strapi/provider-upload-aws-s3": "^4.24.2",
    "@strapi/strapi": "4.24.2",
    "mysql": "^2.18.1",
    "react": "^18.3.1",
    "react-dom": "^18.3.1",
    "react-router-dom": "^5.3.4",
    "styled-components": "^5.3.11"
  }
}
```
config/middlewares.js :- 
```
module.exports = [
  "strapi::errors",
  "strapi::security",
  "strapi::poweredBy",
  {
    name: "strapi::cors",
    config: {
      origin: ["http://localhost:3000"],
    },
  },
  "strapi::logger",
  "strapi::query",
  "strapi::body",
  "strapi::favicon",
  "strapi::public",
];
```
I've been trying different solutions but not getting anywhere. Can someone shed some light on this issue?

**Comment by Garvitpant777** [comment]
can i work on this?

**Comment by EstebanPiacentino** [comment]
Hello!

I'm having the same problem as well in Strapi v4.25.8. This is critical for security reasons.
Is there any ETA for this or a temporary solution?

Many thanks

**Comment by alexandrebodin** [comment]
can you share your server configuration too please ?

**Comment by EstebanPiacentino** [comment]
I'm testing this locally with node v18.19.0 using NODE_ENV as production.

System Info:

- Node.js version: v18.19.0
- NPM version: npm/10.2.3
- Strapi version: 4.25.8
- Database: PostgreSQL
- Operating system: macOS M1
- Is your project Javascript or Typescript: Javascript

strapi package.json:

```
{
  "name": "test",
  "private": true,
  "version": "0.0.0",
  "description": "test back end application",
  "scripts": {
    "postinstall": "npx patch-package",
    "build": "strapi build",
    "develop": "strapi develop",
    "develop-deb": "strapi develop --debug",
    "start": "node node_modules/@strapi/strapi/bin/strapi.js start",
    "strapi": "strapi"
  },
  "dependencies": {
    "@ckeditor/strapi-plugin-ckeditor": "^0.0.13",
    "@strapi/helper-plugin": "^4.25.6",
    "@strapi/plugin-color-picker": "^4.25.6",
    "@strapi/plugin-email": "^4.25.6",
    "@strapi/plugin-i18n": "^4.25.6",
    "@strapi/plugin-users-permissions": "^4.25.6",
    "@strapi/provider-email-nodemailer": "^4.25.6",
    "@strapi/provider-email-sendgrid": "^4.25.6",
    "@strapi/provider-upload-aws-s3": "^4.25.6",
    "@strapi/strapi": "^4.25.6",
    "axios": "^1.7.3",
    "better-sqlite3": "11.1.2",
    "cheerio": "^1.0.0-rc.12",
    "csv-parse": "^5.5.6",
    "dotenv": "^16.4.5",
    "https": "^1.0.0",
    "jwt-decode": "^4.0.0",
    "lodash": "^4.17.21",
    "node-cache": "^5.1.2",
    "npm-check-updates": "^17.0.3",
    "patch-package": "^8.0.0",
    "pg": "^8.12.0",
    "react": "^18.3.1",
    "react-dom": "^18.3.1",
    "react-router-dom": "^5.3.4",
    "rollbar": "^2.26.4",
    "strapi-plugin-populate-deep": "^3.0.1",
    "strapi-plugin-preview-button": "^2.2.2",
    "strapi-provider-email-azure": "^1.0.3",
    "styled-components": "^5.3.11"
  },
  "engines": {
    "node": ">=18.0.0 <=20.x.x",
    "npm": ">=9.0.0 <=10.x.x"
  },
  "license": "MIT"
}
```

config/middlewares.js :

```
module.exports = ({ env }) => {
  return [
    'strapi::logger',
    'strapi::errors',
    'strapi::security',
    {
      name: 'strapi::cors',
      config: {
        headers: '*',
        origin: ['http://0.0.0.0:1338','http://0.0.0.0:5713']
      },
    },
    'strapi::poweredBy',
    'strapi::query',
    'strapi::body',
    'strapi::session',
    'strapi::favicon',
    'strapi::public',
    {
      name: 'strapi::security',
      config: {
        contentSecurityPolicy: {
          directives: {
            'connect-src': ["'self'", 'https:'],
            'img-src': ["'self'", 'data:', 'blob:', 'dl.airtable.com', `${env('AWS_BUCKET', '')}.s3.${env('AWS_REGION', '')}.amazonaws.com`],
            'media-src': ["'self'", 'data:', 'blob:', 'dl.airtable.com', `${env('AWS_BUCKET', '')}.s3.${env('AWS_REGION', '')}.amazonaws.com`],
            upgradeInsecureRequests: null,
          },
        },
      },
    },
    'strapi::compression',
    'global::audit',
    'global::rollbar',
    'global::populate-linked-members'
  ]
}
```

Hope this helps

**Comment by alexandrebodin** [comment]
Sorry I need the config inside `config/server.js` actually :)

**Comment by alexandrebodin** [comment]
Hey

Ok so I understand what's unclear for you, I'll try to explain the best I can 🤞 

When the client is on the same origin (e.g admin served by the server which is the default behavior of strapi) the browser is not sending the `Origin` header because it is the same origin. 

The CORS spec (and that's what can be quite confusing) says that if we do not receive an `Origin` header we should do nothing and allow the request. And this is because it is only a browser thing. So we allow any request without an origin. (e.g a curl request, or a browser request from the same origin)

The validation only occurs when the requesting origin is different from the server origin. To test this you need to create another client and fetch the Strapi api from it and that's when your validation will work. 

In conclusion, there is not way to block same origin requests with cors and previsouly this was because of a bad implementation that it was occuring. 

If you still need to reproduce this behavior you can create a custom middleware for your usecase 👍

**Comment by EstebanPiacentino** [comment]
> Hey
> 
> Ok so I understand what's unclear for you, I'll try to explain the best I can 🤞
> 
> When the client is on the same origin (e.g admin served by the server which is the default behavior of strapi) the browser is not sending the `Origin` header because it is the same origin.
> 
> The CORS spec (and that's what can be quite confusing) says that if we do not receive an `Origin` header we should do nothing and allow the request. And this is because it is only a browser thing. So we allow any request without an origin. (e.g a curl request, or a browser request from the same origin)
> 
> The validation only occurs when the requesting origin is different from the server origin. To test this you need to create another client and fetch the Strapi api from it and that's when your validation will work.
> 
> In conclusion, there is not way to block same origin requests with cors and previsouly this was because of a bad implementation that it was occuring.
> 
> If you still need to reproduce this behavior you can create a custom middleware for your usecase 👍

Amazing! All clear now, the previous behavior was making me think something was wrong :-)
I tried what you said and I can correctly see the CORS working.

Thank you!

**Comment by alexandrebodin** [comment]
Going to close the issue, I think the explanation should help if anyone else asks :)
